Rework of cfn-stack-rename

vamegh opened the pull request:
Hi Ian, 

I came across your tool which is a great idea - I have reworked it quite a bit. I have not as yet properly tested it only tested that things are going through as expected which they all seem to be. 

figured I would raise this pr for you to take a look at. If you are interested in this approach / want to merge it feel free, if you have any suggestions / improvements etc. feel free to comment / let me know. Completely up to you what you do with this. I figured it was only fair to contribute this back into your project - if it is any good.

vamegh committed:
Added initial refactor code - this is work in progress still.

ToDo: Get exports from stack - use this to check what stacks import
the values.

Build a list of resources that cannot be exported - provide option to
migrate them ignoring changesets.

finish the rewrite of index.py into stack_rename.

vamegh committed:
refactored the drift detection code and put it in stack_rename

added comments in stack_rename as to what is required next - this is
mostly done now.

	modified:   libs/aws_handler.py
	modified:   stack_rename

vamegh committed:
Further refactor.

Moved the cfn resource identifiers to a yaml config file - it shouldnt
be hardcoded into a dictionary - easier to update / modify in a yaml
config file.

added update_stack and waiter to aws_handler

slowly going through and refactoring the rest of index.py - before
adding in the functionality I need to it.

	new file:   configs/cloudformation.yaml
	modified:   configs/config.yaml
	modified:   libs/aws_handler.py
	modified:   libs/command_checker.py
	modified:   stack_rename

vamegh committed:
I believe stack_rename should have all current functionality of index.py

The tool has been pretty much refactored.

aws_handler added the following:

create_changeset
exec_changeset
list_imports
list_exports
delete_stack
waiter

Added all original functions - with a twist

it does not bomb out if it encounters resources that dont support
importing instead it lists them - but currently it is not printing out
this information.

the list_imports should list all of the stacks using the exports created
by the original stack (the code still needs to be written still) will
need to go through each stack that imports the resource and change that
specific import to be the actual value of the resource rather than an
"import" statement - theoretically this could be something random - it
may break the service but it will be temporary.

There is a function: resolvePropertyValue which I believe was originally
intended to go through the template and replace all subs and refs into
other values - however this is not used any where. I have not included
it.

No where in the documentation does it state that using a ref for a
resource that doesnt exist cant be done - I think the deployment may
fail for a new resource being built from scratch, but not for an import.
For the time being however I am not tackling this - this should be
addedjust for keeping thing sane.

I believe as we are retaining resources - if they reference
resources that cannot be retained / imported - the resources will exist
but become broken on the next update however those resources will once
again exist so should work again - this is the theory however
cloudformation is insanely dumb. Why would anyone use this ? Terraform
exists and absolutely destroys this mess of a deployment system. All
this code to rename a stack ?? Why is this not available as a function
without all of this mess - also stack drift what a farsical way of
handling this - aws has apis it can query itself to know the state of
the system - why do we have ito manually run this process ? Why is it not
capable of maintaining its own "drift" status ? Just awful, supremely
awful.

	modified:   libs/aws_handler.py
	modified:   stack_rename

vamegh committed:
Further refactor work.

Moved all of the functions from stack_rename to libs/cloudformation.py

Converted cloudformation.yaml config file to cloudformation.json
reformatted cloudformation.yaml as well.

index.py - disabled it from actually doing anything just to check how it
works compared to my changes.

aws_handler fixed a few issues around next tokens - it cannot be a null
value, so making sure it isnt.

command_checker reads cloudformation.json over cloudformation.yaml for
now.

disabled any actual actions from stack_rename it just prints out what it
is doing and the data it is manipulating.

	new file:   configs/cloudformation.json
	modified:   configs/cloudformation.yaml
	modified:   configs/config.yaml
	modified:   index.py
	modified:   libs/aws_handler.py
	new file:   libs/cloudformation.py
	modified:   libs/command_checker.py
	modified:   stack_rename

vamegh committed:
A lot of the base functionality is now working.

It creates 2 different lists based on stuff that is importable and stuff
that isnt, if it isnt importable they are removed from the template
ready for the necessary changeset to be created.

The code checking for drift is actually a bit problematic - not all
resources support drift checking but they may still be importable - this
is something to keep in mind. I think a future version may have to
include a flag to disable / enable drift checking as I actually dont
really need it.

config.yaml: changed back to using cloudformation.yaml - prefer to keep
all configs in yaml format as easier to read.

index.py: added a bunch of debug statements.

libs/aws_handler: some minor cleanups around removed print of client
object and changed cfn template detection to info from warning.

libs/cloudformation:

fixed sanitize_template function - this now copies the template and
removes all keys that are not importable as well as properly outputting
the importable / non-importable stuff - this is still going to be
cleaned up further.

working on sanitize_resources still.

command_checker - changed back to the yaml file for the configuration
data.

stack_rename - added some debugging statements, updated the
sanitize_template call to provide the updated template.

updated all functions further to use the sanitized template provided
back by the sanitize_template function.

	modified:   configs/config.yaml
	modified:   index.py
	modified:   libs/aws_handler.py
	modified:   libs/cloudformation.py
	modified:   libs/command_checker.py
	modified:   stack_rename

vamegh committed:
Updated all of the importable resources config data

using: https://github.com/iann0036/former2/blob/master/util/importable.txt
updated all of the config data so that we have an up to date list of
what can be imported.

added config_convertor.py which generates yaml and json from
libs/cfn_resource_identifier.py.

	new file:   config_convertor.py
	modified:   configs/cloudformation.json
	modified:   configs/cloudformation.yaml
	modified:   libs/cfn_resource_identifiers.py

vamegh committed:
It continues..

updated the cloudformation import configs - the updated list from
former2 didnt have aws iam roles and a few other things - added them
back in.

cloudformation.py - added stack_exports, stacks_importing_exports,
remove_imports_from_stack_templates(not used)

can now list all of the exports and all of the stacks importing them
with a nice big warning about unlinking the stack by removing the
imports from other stacks before continuing.

updated sanitize_template so it generates dicts rather than lists and
now keeps track of supported importable resources, unsupported and
resources that dont support drift.

sanitize_resources - added more debugging to track what its doing /
where it is breaking.

stack_rename - updated this to reflect the changes in cloudformation.py

	modified:   configs/cloudformation.json
	modified:   configs/cloudformation.yaml
	modified:   libs/cfn_resource_identifiers.py
	modified:   libs/cloudformation.py
	modified:   stack_rename

vamegh committed:
Further cleanups.

cloudformation.yaml removed DeploymetId and FilterName - causing issues.

cloudformation.py added verify_action function to add in some
verification at every step of the process.

removed out some of the obvious commented out debug statemenets.

changed the functions to deepcopy the templates before making any
changes - so they always return a copied modified version of the
template.

stack_rename cleaned up comments and commented out code.
implemented the verify_action to each step.

everything should be complete - this needs some further verification
before actually running it and testing to make sure it works.

	modified:   configs/cloudformation.yaml
	modified:   libs/cloudformation.py
	modified:   stack_rename

vamegh committed:
Refactor pretty much complete.

Cleanups - removed pprints changed logging orders to make it less
noisy although at info it is still pretty noisy.

removed a bunch of commented out pprints and removed pprint calls.

	modified:   libs/cloudformation.py
	modified:   stack_rename

vamegh committed:
Further Cleanups

readme - updated to show current state

updated cloudformation.json, cloudformation.yaml and
cfn_resource_identifiers to match each other.

removed index.py

cloudformation.py - updates to sanitize_resourcs to add retain deletion
to all resources

updated requirements.txt

stack_rename fixed issue it should be using sanitized_resource_template
for the change set creation - fixed this.
updated call to sanitize_resources to pass through the
supported_imports.

Added setup.sh and setup.py to make setup / usage easy - documented in
readme.

	modified:   README.md
	renamed:    config_convertor.py -> config_convertor
	modified:   configs/cloudformation.json
	modified:   configs/cloudformation.yaml
	deleted:    index.py
	modified:   libs/cfn_resource_identifiers.py
	modified:   libs/cloudformation.py
	modified:   requirements.txt
	new file:   setup.py
	new file:   setup.sh
	modified:   stack_rename

iann0036 commented:
Hey @vamegh,

Thanks for your efforts. I'm likely going to keep this repo as-is and potentially archive it in the near future as it's not maintained, however if you have a repo you plan on owning for your variation, I'd love to link to it from here.

Let me know if you do.

vamegh committed:
Some bug fixes and added initial support for s3 and deploying using s3 buckets - 50kb local template limitation

vamegh committed:
Added s3 command line support and some minor cleanups.

	modified:   libs/aws_handler.py
	modified:   libs/cloudformation.py
	modified:   libs/command_checker.py
	modified:   libs/command_parser.py
	modified:   stack_rename

vamegh committed:
Further changes to support s3 integration

vamegh committed:
s3 integration should be complete

vamegh committed:
Bug fixes the s3 integration works now

vamegh committed:
updates and cleanups.

updated readme with more information.

aws_handler added some error checking if trying to catch trying to add
retain twice.

cloudformation_gather:
*  added verify_error

* made sure non-driftables only contains resources that supports
  importing but dont support drift.

* sanitize_resources - added functionality to create import_properties
  for importable but non-driftable resources. The logic around
  import_properties is still a bit broken notes added in the file.

* stack_rename rejigged things about a bit. The delete is also awful
  in theory it should be fine - however it can create the changeset, but
  fail on the execute changeset and with the stack deleted no way of
  running things again - its a 1 way trip to cloudformation hell.

	modified:   README.md
	modified:   libs/aws_handler.py
	modified:   libs/cloudformation_gather.py
	modified:   stack_rename

vamegh committed:
Added initial stack state logging / tracking.

The initial state of cfn stacks tracking / logging has been added,
right now a full output of what exactly is being changed should be
stored - there will be a recovery script added - I already have a very
rough version I am using - this will be made generic / usable and
committed later.

	modified:   libs/cloudformation_gather.py
	modified:   libs/command_checker.py
	modified:   libs/file_handler.py
	modified:   stack_rename

vamegh committed:
This commit adds the following updates:

cloudformation.py:
  verify_drift - check if to use expected properties or
  actual_properties, I think lambdas require the expected properties

  del_res_template added as output for sanitize_template so we can
  delete resources while retaining the original stack before finally
  deleting the stack after exec change set has finished.

  sanitize_resources updated to call verify_drift and give the user the
  choice for using which ever properties they prefer.

  fixed import_properties for identifiers with more than 1 identifier to
  get the others from the actual properties which is where they should
  reside.

command_checker:
  added read_state this is a follow on from previous commit to be able
  recover from a failed import.

command_parser:
   added time_stamp again for recovery of failed import

stack_rename:
    using the del_res_template to more safely handle the stacks.

	modified:   libs/cloudformation_gather.py
	modified:   libs/command_checker.py
	modified:   libs/command_parser.py
	modified:   stack_rename

vamegh committed:
Updates as follows:

config_files - added back required importProperties back into config
files.

cloudformation_gather: fixed typo and added more logging.

command_checker: added creation of current_state file.

stack_rename: added s3 support for del_resource_template

	modified:   configs/cloudformation.json
	modified:   configs/cloudformation.yaml
	modified:   libs/cfn_resource_identifiers.py
	modified:   libs/cloudformation_gather.py
	modified:   libs/command_checker.py
	modified:   stack_rename

vamegh committed:
Plan to remove resources before deleting stack failed

vamegh committed:
Added stack_recover

	new file:   libs/cloudformation_recover.py
	modified:   setup.py
	new file:   stack_recover

vamegh committed:
added a method to add dependencies to resources - which doesnt work
properly.

some bug fixes for stack_recover - it does seem to be fully working now.

	modified:   libs/cloudformation_recover.py
	modified:   stack_recover

vamegh committed:
Clean ups - rmeoved all of the commented out code.

Removed all calls/methods that are not used / did not work as expected.

updated some of the comments - this is pretty much feature complete now
and I doubt I will be making any more updates to it.

	modified:   libs/cloudformation_recover.py
	modified:   stack_recover
	modified:   stack_rename

vamegh committed:
stack_rename v0.99beta1

Just to be difficult gave it an awkard version - its actually more like
alpha to be honest.

It works, I have managed to rename a fairly complex stack using it and
the recovery tool - it is not perfect, but then neither is
cloudformation. If you have any say or any control in your organisation
ditch cloudformation and use terraform, seriously cloudformation is by
far the worst cloud deployment tool I have come across.

use this at your own peril - no guarantees.

	modified:   README.md
	modified:   setup.py
	modified:   stack_recover
	modified:   stack_rename

vamegh commented:
Hey sorry for not getting back to you, I was actually very close to abandoning this as it has been a tad bit infuriating :) I have gotten it to the point where I am ok with it. if you want to include it I will leave it here: https://github.com/vamegh/cfn-stack-rename I am an awful maintainer - I don't know if I will get the chance to maintain it - either way do with it as you please :)

vamegh committed:
Ensure the current_state is only modified when current state is completely known - not before

vamegh committed:
A couple of bug fixes.

The len for exports / imports should be greater than 0.

updated write_json it should check to see if data is None explicitly.

Added a few logging.debug statements.

	modified:   libs/cloudformation_gather.py
	modified:   libs/file_handler.py

iann0036 reviewed (approved):
Apologies, thought this was a small change. Re-opening.
